improver/calibration/rainforest_calibration.py

```diff
@@ -126,6 +144,433 @@ def __init__(self, model_config_dict: dict, threads: int):
             ]
 
+    def _check_num_features(self, features: CubeList) -> None:
+        """Check that the correct number of features has been passed into the model.
+
+        Args:
+            features:
+                Cubelist containing feature variables.
+        """
+        from lightgbm import Booster
+
+        sample_tree_model = self.tree_models[0]
+
+        if isinstance(sample_tree_model, Booster):
+            expected_num_features = sample_tree_model.num_feature()
+        elif self.treelite_enabled:
+            from treelite_runtime import Predictor
+
+            if isinstance(sample_tree_model, Predictor):
+                expected_num_features = sample_tree_model.num_feature
+
+        if expected_num_features != len(features):
+            raise ValueError(
+                "Number of expected features does not match number of feature cubes."
+            )
+
+    def _align_feature_variables(
+        self, feature_cubes: CubeList, forecast_cube: Cube
+    ) -> Tuple[CubeList, Cube]:
+        """Ensure that feature cubes have consistent dimension coordinates. If realization
+        dimension present in any cube, all cubes lacking this dimension will have realization
+        dimension added and broadcast along this new dimension.
+
+        This situation occurs when derived fields (such as accumulated solar radiation)
+        are used as predictors. As these fields do not contain a realization dimension,
+        they must be broadcast to match the NWP fields that do contain realization, so that
+        all features have consistent shape.
+
+        In the case of deterministic models (those without a realization dimension), a
+        realization dimension is added to allow consistent behaviour between ensemble and
+        deterministic models.
+
+        Args:
+            feature_cubes:
+                Cubelist containing feature variables to align.
+            forecast_cube:
+                Cube containing the forecast variable to align.
+
+        Returns:
+            - feature_cubes with realization coordinate added to each cube if absent
+            - forecast_cube with realization coordinate added if absent
+
+        Raises:
+            ValueError:
+                if feature/forecast variables have inconsistent dimension coordinates
+                (excluding realization dimension), or if feature/forecast variables have
+                different length realization coordinate over cubes containing a realization
+                coordinate.
+        """
+        combined_cubes = CubeList(list([*feature_cubes, forecast_cube]))
+
+        # Compare feature cube coordinates, raise error if dim-coords don't match
+        compare_feature_coords = compare_coords(
+            combined_cubes, ignored_coords=["realization"]
+        )
+        misaligned_dim_coords = [
+            coord_info["coord"]
+            for misaligned_coords in compare_feature_coords
```

Comment:
Good suggestion. Looking back over the double loop, it took me sometime to digest, so that's a good indication that this needs simplifying. I'll make the change as you've suggested, and I'll also make a similar change to the check between realization coordinates.